// @hashgraph/sdk/lib/nft/NftInfoQuery.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _Query = _interopRequireWildcard(require("../query/Query.cjs"));

var _NftId = _interopRequireDefault(require("./NftId.cjs"));

var _TokenNftInfo = _interopRequireDefault(require("./TokenNftInfo.cjs"));

var _Hbar = _interopRequireDefault(require("../Hbar.cjs"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * @namespace proto
 * @typedef {import("@hashgraph/proto").IQuery} proto.IQuery
 * @typedef {import("@hashgraph/proto").IQueryHeader} proto.IQueryHeader
 * @typedef {import("@hashgraph/proto").IResponse} proto.IResponse
 * @typedef {import("@hashgraph/proto").ITokenNftInfo} proto.ITokenNftInfo
 * @typedef {import("@hashgraph/proto").IResponseHeader} proto.IResponseHeader
 * @typedef {import("@hashgraph/proto").ITokenGetNftInfoQuery} proto.ITokenGetNftInfoQuery
 * @typedef {import("@hashgraph/proto").ITokenGetNftInfoResponse} proto.ITokenGetNftInfoResponse
 */

/**
 * @typedef {import("../channel/Channel.js").default} Channel
 */

/**
 * @augments {Query<TokenInfo>}
 */
class NftInfoQuery extends _Query.default {
  /**
   * @param {object} properties
   * @param {NftId | string} [properties.nftId]
   */
  constructor(properties = {}) {
    super();
    /**
     * @private
     * @type {?NftId}
     */

    this._nftId = null;

    if (properties.nftId != null) {
      this.setNftId(properties.nftId);
    }
  }
  /**
   * @internal
   * @param {proto.IQuery} query
   * @returns {NftInfoQuery}
   */


  static _fromProtobuf(query) {
    const info =
    /** @type {proto.ITokenGetNftInfoQuery} */
    query.tokenGetNftInfo;
    return new NftInfoQuery({
      nftId: info.nftID != null ? _NftId.default._fromProtobuf(info.nftID) : undefined
    });
  }
  /**
   * @returns {?NftId}
   */


  get nftId() {
    return this._nftId;
  }
  /**
   * Set the token ID for which the info is being requested.
   *
   * @param {NftId | string} nftId
   * @returns {NftInfoQuery}
   */


  setNftId(nftId) {
    this._nftId = typeof nftId === "string" ? _NftId.default.fromString(nftId) : _NftId.default._fromProtobuf(nftId._toProtobuf());
    return this;
  }
  /**
   * @override
   * @param {import("../client/Client.js").default<Channel, *>} client
   * @returns {Promise<Hbar>}
   */


  async getCost(client) {
    let cost = await super.getCost(client);

    if (cost.toTinybars().greaterThan(25)) {
      return cost;
    } else {
      return _Hbar.default.fromTinybars(25);
    }
  }
  /**
   * @override
   * @internal
   * @param {Channel} channel
   * @param {proto.IQuery} request
   * @returns {Promise<proto.IResponse>}
   */


  _execute(channel, request) {
    return channel.token.getTokenNftInfo(request);
  }
  /**
   * @override
   * @internal
   * @param {proto.IResponse} response
   * @returns {proto.IResponseHeader}
   */


  _mapResponseHeader(response) {
    const tokenGetNftInfo =
    /** @type {proto.ITokenGetNftInfoResponse} */
    response.tokenGetNftInfo;
    return (
      /** @type {proto.IResponseHeader} */
      tokenGetNftInfo.header
    );
  }
  /**
   * @override
   * @override
   * @internal
   * @param {proto.IResponse} response
   * @returns {Promise<TokenInfo>}
   */


  _mapResponse(response) {
    const info =
    /** @type {proto.ITokenGetNftInfoResponse} */
    response.tokenGetNftInfo;
    return Promise.resolve(_TokenNftInfo.default._fromProtobuf(
    /** @type {proto.ITokenNftInfo} */
    info.nft));
  }
  /**
   * @override
   * @internal
   * @param {proto.IQueryHeader} header
   * @returns {proto.IQuery}
   */


  _onMakeRequest(header) {
    return {
      tokenGetNftInfo: {
        header,
        nftID: this._nftId != null ? this._nftId._toProtobuf() : null
      }
    };
  }

} // eslint-disable-next-line @typescript-eslint/unbound-method


exports.default = NftInfoQuery;

_Query.QUERY_REGISTRY.set("tokenGetNftInfo", NftInfoQuery._fromProtobuf);
